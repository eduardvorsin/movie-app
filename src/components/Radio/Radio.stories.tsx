import type { Meta, StoryObj } from '@storybook/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { action } from '@storybook/addon-actions';
import Radio, { Props } from './Radio';
import { ChangeEvent, useState } from 'react';
import Button from '@/components/Button/Button';

const meta: Meta<typeof Radio> = {
	title: 'UI/Radio',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for radio buttons'
		},
		testId: {
			description: 'Id for testing the component',
		},
		label: {
			description: 'Text description for the input',
		},
		value: {
			description: 'Value for the radio button',
		},
		name: {
			description: 'Name for the input',
		},
		id: {
			description: 'Radio button id',
		},
		isChecked: {
			description: 'Determines whether the radio button is marked or not',
		},
		isDisabled: {
			description: 'Disabling the radio button makes it non-functional',
		},
		isInvalid: {
			description: 'Visually highlights the radio button if it does not pass validation',
		},
		isRequired: {
			description: 'Marks the radio button as mandatory',
		},
		onFocus: {
			description: 'The callback is executed when the focus hits the input',
			control: false,
		},
		onBlur: {
			description: 'The callback is executed when the input loses focus',
			control: false,
		},
		onChange: {
			description: 'A callback is called when the radio button changes state from marked to unmarked and vice versa',
			control: false,
		},
	},
	component: Radio,
	parameters: {
		docs: {
			description: {
				component: 'A Radio is an individual element within a RadioGroup. It\'s highly advised to utilize Radio elements within a RadioGroup for proper functionality. Nonetheless, you can also use them independently: if multiple Radio items share the same name attribute, they will be considered as belonging to the same group.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof Radio>;

const RadioWithHooks = (props: Omit<Props, 'onChange' | 'isChecked'>) => {
	const [isChecked, setIsChecked] = useState<boolean>(false);

	const setChecked = (e: ChangeEvent<HTMLInputElement>): void => {
		setIsChecked(true);
		action('Changed')(e);
	};

	const clearChecked = (): void => setIsChecked(false);

	return (
		<div className='flex flex-col items-start'>
			<Radio
				{...props}
				isChecked={isChecked}
				onChange={setChecked}
			/>

			<Button
				className='mt-4'
				onClick={clearChecked}
			>
				Clear
			</Button>
		</div>
	)
};

export const Default: Story = {
	args: {
		id: 'option',
		name: 'option',
		label: 'option',
		value: 'value',
	},
	parameters: {
		docs: {
			description: {
				story: 'Default state of the radio button',
			},
		},
	},
	render: (args) => (<RadioWithHooks {...args} />),
};

export const Disabled: Story = {
	args: {
		id: 'disabled',
		name: 'disabled',
		label: 'disabled',
		value: 'value',
		isDisabled: true,
	},
	parameters: {
		docs: {
			description: {
				story: 'Implement this for actions that are presently inaccessible. The user interface should clearly communicate why the input is inactive and provide guidance on the necessary steps to activate it.',
			},
		},
	},
	render: (args) => (<RadioWithHooks {...args} />),
};

export const Invalid: Story = {
	args: {
		id: 'invalid',
		name: 'invalid',
		label: 'invalid',
		value: 'value',
		isInvalid: true,
	},
	parameters: {
		docs: {
			description: {
				story: 'It is utilized to visually emphasize the input by turning it red when it fails validation.',
			},
		},
	},
	render: (args) => (<RadioWithHooks {...args} />),
};

export const EventCallbacks: Story = {
	argTypes: {
		onFocus: {
			action: 'Focused',
		},
		onBlur: {
			action: 'Blured',
		},
	},
	args: {
		id: 'event-callbacks',
		name: 'event-callbacks',
		label: 'focus, blur event',
		value: 'value',
	},
	parameters: {
		docs: {
			description: {
				story: 'The callback function can be provided to the onFocus, onBlur event.',
			},
		},
	},
	render: (args) => (<RadioWithHooks {...args} />),
};