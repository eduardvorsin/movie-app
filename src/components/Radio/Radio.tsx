'use client';
import { ChangeEventHandler, FocusEventHandler } from 'react';

type Props = {
	className?: string,
	id?: string,
	value?: string,
	testId?: string,
	name?: string,
	label?: string,
	isChecked?: boolean,
	isDisabled?: boolean,
	isRequired?: boolean,
	isInvalid?: boolean,
	onChange?: ChangeEventHandler<HTMLInputElement>,
	onFocus?: FocusEventHandler<HTMLInputElement>,
	onBlur?: FocusEventHandler<HTMLInputElement>,
};

export default function Radio({
	className,
	id,
	value,
	name,
	label,
	testId,
	isDisabled,
	isChecked,
	isRequired,
	isInvalid,
	onChange,
	onFocus,
	onBlur,
}: Props) {
	const labelClasses = [
		'inline-flex py-1 pl-5 items-center',
		isDisabled ? 'opacity-disabled cursor-not-allowed' : 'cursor-pointer',
		className,
	].join(' ');

	return (
		<label
			className={labelClasses}
			data-testid={testId}
		>
			<input
				className='appearance-none absolute peer'
				type='radio'
				id={id}
				name={name}
				disabled={isDisabled}
				onChange={onChange}
				onFocus={onFocus}
				onBlur={onBlur}
				value={value}
				checked={isChecked}
				required={isRequired}
				data-invalid={isInvalid}
			/>
			<span
				className='relative w-5 h-5 -ml-5 rounded-full border-2 border-neutral-500 peer-checked:border-blue-500 peer-focus:border-blue-500 peer-data-[invalid]:border-red-600 transition-colors duration-100 ease-linear dark:peer-data-[invalid]:border-red-500  dark:peer-focus:border-blue-300 dark:peer-checked:border-blue-300 dark:border-dark-neutral-800 after:w-2 after:h-2 after:rounded-full after:absolute peer-checked:after:bg-blue-700 dark:peer-checked:after:bg-blue-400 after:top-1/2 after:left-1/2 after:-translate-x-1/2 after:-translate-y-1/2'
			/>
			<span
				className='ml-2 text-100 text-dark-neutral-0 dark:text-neutral-0'
			>
				{label}
			</span>
		</label>
	);
};