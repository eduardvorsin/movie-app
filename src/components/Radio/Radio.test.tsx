import { render, screen } from '@testing-library/react';
import Radio from './Radio';
import userEvent from '@testing-library/user-event';

describe('Radio tests', () => {
	it('is rendered correctly', () => {
		render(
			<Radio
				id='test-radio'
				name='test-radio'
				label='test-radio'
				value=''
				onChange={() => { }}
				isChecked={false}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('radio')).toBeInTheDocument();
	});

	it('when you click on the radio button, the mock function is called', async () => {
		const mockFn = jest.fn();
		const user = userEvent.setup()
		render(
			<Radio
				id='test-radio'
				name='test-radio'
				label='test-radio'
				value=''
				onChange={mockFn}
				isChecked={false}
			/>
		);

		await user.click(screen.getByRole<HTMLInputElement>('radio'));

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('when the focus hits the radio button, the mock function is called', async () => {
		const mockFn = jest.fn();
		const user = userEvent.setup()
		render(
			<Radio
				id='test-radio'
				name='test-radio'
				label='test-radio'
				value=''
				onChange={() => { }}
				onFocus={mockFn}
				isChecked={false}
			/>
		);

		await user.tab();

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('if the focus is lost, the mock function is called from the radio button', async () => {
		const mockFn = jest.fn();
		const user = userEvent.setup()
		render(
			<Radio
				id='test-radio'
				name='test-radio'
				label='test-radio'
				value=''
				onChange={() => { }}
				onFocus={mockFn}
				isChecked={false}
			/>
		);

		await user.tab();
		await user.click(document.body);

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('if isDisabled is true then the radio button becomes inactive', async () => {
		const mockFn = jest.fn();
		const user = userEvent.setup()
		render(
			<Radio
				id='test-radio'
				name='test-radio'
				label='test-radio'
				value=''
				isDisabled
				onChange={mockFn}
				isChecked={false}
			/>
		);

		await user.click(screen.getByRole<HTMLInputElement>('radio'));

		expect(mockFn).toHaveBeenCalledTimes(0);
	});
	it('is a basic snapshot', () => {
		render(
			<Radio
				id='test-radio'
				name='test-radio'
				label='test-radio'
				value=''
				onChange={() => { }}
				isChecked={false}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('radio')).toMatchSnapshot();
	});

	it('is a snapshot with isDisabled set to true', () => {
		render(
			<Radio
				id='test-radio'
				name='test-radio'
				label='test-radio'
				isDisabled
				value=''
				onChange={() => { }}
				isChecked={false}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('radio')).toMatchSnapshot();
	});

	it('is a snapshot with IsInvalid equal to true', () => {
		render(
			<Radio
				id='test-radio'
				name='test-radio'
				label='test-radio'
				isInvalid
				value=''
				onChange={() => { }}
				isChecked={false}
			/>
		);

		expect(screen.getByRole<HTMLInputElement>('radio')).toMatchSnapshot();
	});
});