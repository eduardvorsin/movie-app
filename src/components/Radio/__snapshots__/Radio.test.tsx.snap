// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio tests is a basic snapshot 1`] = `
<input
  class="appearance-none absolute peer"
  id="test-radio"
  name="test-radio"
  type="radio"
  value=""
/>
`;

exports[`Radio tests is a snapshot with IsInvalid equal to true 1`] = `
<input
  class="appearance-none absolute peer"
  data-invalid="true"
  id="test-radio"
  name="test-radio"
  type="radio"
  value=""
/>
`;

exports[`Radio tests is a snapshot with isDisabled set to true 1`] = `
<input
  class="appearance-none absolute peer"
  disabled=""
  id="test-radio"
  name="test-radio"
  type="radio"
  value=""
/>
`;
